include/fiction/technology/sidb_cluster_hierarchy.hpp

```diff
@@ -0,0 +1,1080 @@
+ */
+enum class bound_direction : uint8_t
+{
+    LOWER = 0,
+    UPPER
+};
+/**
+ * The respective *top* elements of the lower semi-lattices in which our potential bounds live, ie., the respective
+ * elements of *most* information. This means that any meet with a top element (weakly) reduces the information, and
+ * thus any meet computed in iteration through binary application may start out with this element.
+ *
+ * @tparam bound Potential bound domain to return the element of most information of.
+ * @return The element of most information respective to the potential bound domain.
+ */
+template <bound_direction bound>
+static constexpr inline double potential_bound_top() noexcept
+{
+    if constexpr (bound == bound_direction::LOWER)
+    {
+        return std::numeric_limits<double>::infinity();
+    }
+    else if constexpr (bound == bound_direction::UPPER)
+    {
+        return -std::numeric_limits<double>::infinity();
+    }
+}
+/**
+ * This function computes a binary meet, overwriting the first argument with the result. It takes the minimum in case of
+ * a lower bound, and a maximum in case of an upper bound, each corresponding to the minimal information common to the
+ * arguments.
+ *
+ * @tparam bound The potential bound domain which defines the implementation of the meet.
+ * @param a First potential bound which is overwritten with the result of the meet.
+ * @param b Second potential bound.
+ */
+template <bound_direction bound>
+static constexpr inline void take_meet_of_potential_bounds(double& a, const double b) noexcept
+{
+    if constexpr (bound == bound_direction::LOWER)
+    {
+        a = std::min(a, b);
+    }
+    else if constexpr (bound == bound_direction::UPPER)
+    {
+        a = std::max(a, b);
+    }
+}
+/**
+ * Forward declaration. Required for compilation due to the mutually recursive structure in this file.
+ */
+static inline uint64_t get_singleton_sidb_ix(const sidb_cluster_ptr& c) noexcept;
+/**
+ * Forward declaration. Required for compilation due to the mutually recursive structure in this file.
+ */
+static inline uint64_t get_unique_cluster_id(const sidb_cluster_ptr& c) noexcept;
+/**
+ * This defines the type for a store in which the bounds on the local potential for an SiDB (index) may be stored. For
+ * the *Ground State Space* algorithm, this is used to keep track of the bounds on the fraction of local potential that
+ * is projected on the SiDB from within a cluster, though during *ClusterComplete* simulation, this is used as a dynamic
+ * store in which also external potentials are accumulated.
+ */
+#ifdef DEBUG_SIDB_CLUSTER_HIERARCHY
+using intra_cluster_potential_bounds = std::map<uint64_t, std::array<double, 2>>;
+#else
+using intra_cluster_potential_bounds = phmap::flat_hash_map<uint64_t, std::array<double, 2>>;
+#endif
+/**
+ * A cluster state is a projector state paired with a potential bound store in which local potentials are stored for
```

Comment:
Do you or can you explain somewhere what a projector state is?

Reply:
It's just that :p simply a specialised pair that now gets a name with the nice `pst` abbreviation for variables that respects the semantics of the data in the pair

Reply:
I thought this commented om the projector state. But anyhow my comment applies to the projector state but exactly in the same way it also applies here. These datastructures are just specialised pairs